RFC: Establish concept of a computing device

oschulz opened the pull request:
CUDA.jl, oneAPI.jl, etc. all provide a `...Device` type, but without a common supertype.

Likewise, our GPU packages all provide functionality to get the device that a given array lives on, but each defines it's own function for it. The latter was partially addressed in (JuliaGPU/KernelAbstractions.jl#269) but it's not an elegant solution (all Requires-based) and KernelAbstractions is a heavy dependency. This makes it tricky to address issues like JuliaArrays/ElasticArrays.jl#44. Some of the code mentioned in JuliaGPU/GPUArrays.jl#409 could also lighten it's dependencies with common "get-device" functionality (LinearSolve.jl, for example, seems to need GPUArray.jl only for a [`b isa GPUArrays.AbstractGPUArray`](https://github.com/SciML/LinearSolve.jl/blob/8e31fd75dd10f478fca74fd46477f7b342590ab9/src/default.jl#L12), similar for [DiffEqSensitivity.jl](https://github.com/SciML/DiffEqSensitivity.jl/blob/50fff85abca5452a90b155f8eb9833498d5b7aa4/src/concrete_solve.jl#L45)

This PR and supporting PRs for CUDA.jl, AMDGPU.jl, oneAPI.jl and KernelAbstractions.jl attempt to establish a common supertype for computing devices, and support for

* `get_computing_device(x)::AbstractComputingDevice`: Get the device `x` lives on, not limited to arrays (could e.g. be a whole ML model)

* `Adapt.adapt_storage(dev, x)`: Move `x` to device `dev`.

* `Sys.total_memory(dev)`: Get the total memory on `dev`

* `Sys.free_memory(dev)`: Get the free memory on `dev`

I think this will make it much easier to write generic device-independent code:

* Being able to query if data lives on a CPU or GPU without taking on heavy dependencies should comes in useful in many packages.

* Writing `adapt(CuDevice(n), x)` as an alternative to `adapt(CuArray, x)` seems very natural (esp. in multi-GPU scenarios): It corresponds to the user saying "let's run it on that GPU" instead of "with a different array type".

* Having the ability to query total and available memory can help with finding the right data chunk sizes before sending data to a device with `adapt(dev, data_partition)`.

This PR defines `AbstractComputingDevice`, `AbstractGPUDevice` and implements `GPUDevices`. It's very little code, there should be no load-time impact.


CC @vchuravy, @maleadt, @jpsamaroo, @ChrisRackauckas, @findmyway

Status:

* no tests yet while waiting for design comments from package maintainers

* CPU: functional

* CUDA: functional with with JuliaGPU/CUDA.jl#1520 (`CuDevice <: Adapt.AbstractGPUDevices`)

* AMDGPU: needs expert advice, draft is here: JuliaGPU/AMDGPU.jl#233

* oneAPI: functional (but not complete) with JuliaGPU/oneAPI.jl#185 (`ZeDevice <: Adapt.AbstractGPUDevice`)

* KernelAbstractions: funcional on CUDA with JuliaGPU/KernelAbstractions.jl#297 (replace `KernelAbstractions.Device` with `Adapt.AbstractComputingDevice`)

ChrisRackauckas commented on src/computedevs.jl:
```diff
@@ -0,0 +1,80 @@
+
+Supertype for GPU computing devices.
+"""
+abstract type AbstractGPUDevice <: AbstractComputingDevice end
```
Maybe have an abstract accelerator device, so it covers all non-CPU like TPU, IPU, etc.

oschulz replied:
> Maybe have an abstract accelerator device, so it covers all non-CPU like TPU, IPU, etc.

Yes, I had hope that this might extend to other compute devices types as well one day (FPGA, ...). Do we need to have an `AbtractComputingAccelerator` or `AbstractNonCPUDevice`? Or should difference kinds of non-CPU devices be direct subtypes of `AbstractComputingDevice`, on the same level as `CPUDevice` (so one would check `!(dev isa CPUDevice)`)?

oschulz replied:
Thanks @AriMKatz , good idea! We definitely want to make sure this is done in a way that everyone can utilize.

oschulz replied:
> Maybe instead of ComputingDevice we call it ComputeUnit? We are moving towards heterogeneous system in general and they might not be separated devices

I'm renaming it accordingly.

oschulz replied:
> Maybe have an abstract accelerator device

I'm adding AbstractComputeAccelerator into the type hierarchy.

vchuravy commented:
@tkf does this jive with what you need for Loops/Executor?

vchuravy commented:
Maybe instead of ComputingDevice we call it ComputeUnit? We are moving towards heterogeneous system in general and they might not be separated devices

oschulz commented:
> Maybe instead of ComputingDevice we call it ComputeUnit? We are moving towards heterogeneous system in general and they might not be separated devices

Sure, absolutely!

Since this touches several PRs maybe I should wait for feedback on the general idea from @maleadt and @jpsamaroo before doing renames and so on?

tkf commented on src/computedevs.jl:
```diff
@@ -0,0 +1,80 @@
+# ToDo: Utilize `ArrayInterfaceCore.buffer(A)`? Would require Adapt to depend
+# on ArrayInterfaceCore.
+
+@generated function get_computing_device(x)
+    impl = :(begin dev_0 = ComputingDeviceIndependent() end)
+    append!(impl.args, [:($(Symbol(:dev_, i)) = select_computing_device(get_computing_device(getfield(x, $i)), $(Symbol(:dev_, i-1)))) for i in 1:fieldcount(x)])
+    push!(impl.args, :(return $(Symbol(:dev_, fieldcount(x)))))
+    impl
+end
```
I would avoid this if I were doing this because:

* It doesn't handle cycles. (Cycles can be created for mutable structs. Recursing only for immutable is a bad idea since mutable vs immutable struct is an implementation detail.)
* There is nothing wrong with a given collection object containing references to "remote" memories. For example, it's possible to have a "CPU array" that is backed up by `Base.Array` but refers to a `CuArray` from which data is copied upon certain events. The data access happens purely through `Base.Array`.

If I were doing this, I would define a fold over "relevant data source objects" in a manner generalizing `Adapt.adapt_structure`.

oschulz replied:
Good point about the loops. I'm also not perfectly happy with this magic, but on the other hand it's unreasonable to expect struct maintainers to implement a large number of methods to make their structures GPU+ML capable. They already have to implement `Adapt.adapt_storage`, Functors.jl and possibly ConstructionBase.jl support, etc ... that's another consensus API we're currently lacking in my opinion, semantic struct tree introspection, de- and re-composition.

> If I were doing this, I would define a fold over "relevant data source objects" in a manner generalizing Adapt.adapt_structure.

I guess there we enter Functors & friends territory ... it might be good is some key people from the array ecosystem would join the discussion in JuliaObjects/ConstructionBase.jl#53 (relevant parts at the end). I feel there's a lot of unused synergy potential and several parallel incompatible solutions in this area currently.

For now, if we want to avoid generated function magic in the PR, we could use `ArrayInterfaceCore.buffer` (which defaults to `Base.parent`). ArrayInterfaceCore.jl is now very lightweight, so I think it could be an acceptable dependency for Adapt.jl (CC @ChrisRackauckas ). So we could do something like

```julia
function get_computing_device(x::AbstractArray)
    buf = buffer(x)
    buf === x && throw(ArgumentError("Can't get buffer of object"))
    get_computing_device(buf)
end
```

oschulz replied:
Or, with a bottom value instead of an exception (see suggestion of @tkf below):

```julia
function get_computing_device(x::AbstractArray)
    buf = ArrayInterfaceCore.buffer(x)
    if typeof(buf) != typeof(x)
        get_computing_device(buf)
    else
        UnknownComputingDevice(x)
    end
end
```

ToucheSir replied:
@tkf I prototyped such a generic fold at https://github.com/FluxML/Functors.jl/pull/32 (and previously https://github.com/Flfolding and uxML/Functors.jl/pull/27). Part of the tension I ran into is that while we may see flattening as a special case of folding/`fmap`ping, the existing Julia ecosystem tends to see it the other way around (traversal happens after flattening).

For this particular PR, I agree the bottom value is the safest option. After all, what would `get_computing_device` return for a NN with mixed CPU and GPU array parameters otherwise?

oschulz replied:
> After all, what would get_computing_device return for a NN with mixed CPU and GPU array parameters otherwise?

I think it should return "incompatible devices"/"can't find common device".

oschulz replied:
> It doesn't handle cycles. 

Cycles are taken care of now.

> There is nothing wrong with a given collection object containing references to "remote" memories. For example, it's possible to have a "CPU array" that is backed up by Base.Array but refers to a CuArray from which data is copied upon certain events. [...]  I would define a fold over "relevant data source objects"

I would say such special types would specialize the new `get_compute_unit_impl`. A generic mechanism to get the "relevant data source objects" of an objects would be nice, of course, but we don't have an ecosystem-wide solution yet I think? When we get one though, we can just use it in the generic `get_compute_unit_impl`.

oschulz replied:
I'll mark this resolved, since I've changed a lot of things, but please feel free to reopen.

tkf commented on src/computedevs.jl:
```diff
@@ -0,0 +1,80 @@
+select_computing_device(a::CPUDevice, ::CPUDevice) = a
+select_computing_device(a::CPUDevice, b::AbstractGPUDevice) = a
+select_computing_device(a::AbstractGPUDevice, b::CPUDevice) = b
+select_computing_device(a::AbstractGPUDevice, b::AbstractGPUDevice) = (a === b) ? a : throw(_incompatible_devs)
```
Note that throwing a nonspecific exception on error makes it impossible to implement an efficient fallback safely. It would be more composable if it were returning some kind of value ("bottom") representing that an appropriate device is undecidable.

oschulz replied:
Oh, yes, I like it - we recently took that approach in InverseFunctions and ChangesOfVariables. If the generated function (see above) goes, we may not need `select_computing_device` at all anymore, at least for now, but we could use a bottom value in the default implementation of `get_computing_device` (see comment by @tkf  above).

oschulz replied:
I'm adding a bottom value

tkf commented on src/computedevs.jl:
```diff
@@ -0,0 +1,80 @@
+
+const _incompatible_devs = ArgumentError("Incompatible computing devices")
+
+select_computing_device(a::ComputingDeviceIndependent, ::ComputingDeviceIndependent) = a
+select_computing_device(a::ComputingDeviceIndependent, b::AbstractComputingDevice) = b
+select_computing_device(a::AbstractComputingDevice, b::ComputingDeviceIndependent) = a
+
+select_computing_device(a::CPUDevice, ::CPUDevice) = a
+select_computing_device(a::CPUDevice, b::AbstractGPUDevice) = a
+select_computing_device(a::AbstractGPUDevice, b::CPUDevice) = b
```
I'd note that this mechanism has a potential extensibility problem (e.g., method ambiguity resolution and associativity/commutativity guarantee). To design this kind of mechanism, I'd look into how it is handled in the broadcast style selection and type promotion rules.

oschulz replied:
This was mainly intended as an internal method to support the "magical" generated default method of `get_computing_device`. If we take that out, I'd suggest to remove `select_computing_device` completely for now.

oschulz commented:
@maleadt and @jpsamaroo do you have some initial feedback? And sorry for the state that the AMDGPU part of this is still in @jpsamaroo, I'll need a few pointers on that. :-)

oschulz commented:
> @maleadt: This seems sensible to me, but I don't understand why it belongs in Adapt.jl.

It seemed to be a natural place, both semantically and dependency-wise.

Semantically, adapt deals with moving storage between devices (at least that's what we use it for, mostly), so it seems natural to have a concept of what a computing device is in here.

Dependency-wise, pretty much all code that will need to define/use `get_computing_device` depends on Adapt already. We could create a super-lightweight package `ComputingDevices.jl` - but Adapt.jl would need to depend on it, so we can define `adapt_storage(::CPUDevice, x)`. I'd be fine with that, but I thought people might prefer not to add such a package since Adapt.jl is already super-lightweight, and code that would need ComputingDevices.jl would very likely need Adapt.jl as well anyway.

> Now, if the proposed device identification were based on the existing Adapt.jl infrastructure (adapt_structure/adapt_storage

Well I with having `adapt_...(dev, x)` (and I think that would be very good to have, see above) it's already integrated, in a way. And even if we only have a simple default implementation of `get_computing_device` that uses `parent`/`buffer` for now, it would certainly be good to expand in the directions that @tkf  suggested. But I think that can be done as a second step, and then benefit Adapt as a whole.

jpsamaroo commented on src/computedevs.jl:
```diff
@@ -0,0 +1,80 @@
+"""
+    abstract type AbstractComputingDevice
+
+Supertype for CPU and GPU computing devices.
```
```suggestion
Supertype for arbitrary computing devices (CPU, GPU, etc.).
```

jpsamaroo commented on src/computedevs.jl:
```diff
@@ -0,0 +1,80 @@
+`CPUDevice()` is the default CPU device.
+"""
+struct CPUDevice <: AbstractComputingDevice end
+export CPUDevice
```
Do we need all of these as exports? "CPUDevice" is pretty non-specific as a name, and could clash with some other package that is aware of computing devices (nothing comes to mind, though).

vchuravy replied:
Oceanigans has his own hierarchy as an example. @glwagner

oschulz replied:
> Do we need all of these as exports? "CPUDevice" is pretty non-specific as a name, and could clash with some other package that is aware of computing devices (nothing comes to mind, though).

I didn't find any package that uses `CPUDevice` currently. I'm always for avoiding name clashes - but since the name is "free": The advantage of exporting this would be to encourage packages to use this new common infrastructure instead of rolling their own, so to speak. :-)

oschulz replied:
I'll mark this resolved since a lot has changed in the PR, please feel free to reopen.

jpsamaroo commented on src/computedevs.jl:
```diff
@@ -0,0 +1,80 @@
+
+const _incompatible_devs = ArgumentError("Incompatible computing devices")
+
+select_computing_device(a::ComputingDeviceIndependent, ::ComputingDeviceIndependent) = a
```
I vote for a name like `promote_computing_device`. "Select" makes it sound like you're doing some imperative action to enable some device, whereas it's really just determining promotion rules.

oschulz replied:
> I vote for a name like `promote_computing_device`

I actually thought about it in that term as well.

oschulz replied:
I was tempted to overload `Base.promote` initially, but then thought that would be stretching it too far. :-) I like `promote_computing_device`.

oschulz replied:
Renaming to promote_computing_device

jpsamaroo reviewed (commented):
Looks good to me generally, aside from existing comments by others. I agree with @maleadt that some part of this could possibly go in GPUArrays, but having it in a more lightweight package instead is definitely appealing.

oschulz commented:
> some part of this could possibly go in GPUArrays, but having it in a more lightweight package instead is definitely appealing

Yes, one important motivation here is to enable generic code to be aware of devices without taking on a heavy dependency like GPUArrays.

oschulz commented:
@vchuravy , @ChrisRackauckas , @tkf , @maleadt , @jpsamaroo thanks for all the inital feedback!

I've updated this PR and tried to include your suggestions - since a lot has changed I've marked the discussions above as resolved, but please do reopen them if issues haven't been addressed in the PR changes and the remarks below:

* Renamed AbstractComputingDevice to AbstractComputeUnit (suggested by @vchuravy) and renamed `get_compute_device` to `get_compute_unit` accordingly.

* Added AbstractComputeAccelerator (suggested by @ChrisRackauckas) in between `AbstractComputeUnit` and `AbstractGPUDevice` (is the type hierarchy too deep now?).

* Using bottom values if compute device can't be determined or unified/merged (suggested by @tkf)

* Renamed `select_computing_device`  (suggestions by @tkf and @jpsamaroo) - it's called `merge_compute_units` now. It can actually be more of a combination than a promotion, I think - for example, multiple CUDA devices in the same box could be merged to a `MultiCUDASystem([0,1,2,3])` or so in the future. Currently, `get_compute_unit` and `merge_compute_units` will return `MixedComputeSystem()` if different CPU/GPU devices are involved.

* The generic implementation of `get_compute_unit` defends against reference cycles (pointed out by @tkf) now.

I think having an automatic recursive compute unit resolution will be important so that `get_compute_unit_impl` won't have to be specialized for most types - and we should support closures. This works:


```julia
julia> mllayer = let A = cu(rand(5,5)), b = cu(rand(5)), f = x -> x < 0 ? zero(x) : x
           x -> f.(A * b .+ b)
       end;

julia> x = cu(rand(5));

julia> get_compute_unit((mllayer, x))
CuDevice(0): NVIDIA GeForce
```

@tkf suggested a "fold over 'relevant data source objects' in a manner generalizing Adapt.adapt_structure". The generated default `get_compute_unit_impl` does such a fold, but since it doesn't need to reconstruct objects it's not limited to tuples and functions like the current `Adapt.adapt_structure`. It currently uses all fields - to my knowledge we don't have a widely supported way to get the "relevant data source objects" yet (that would be neat, though!). I suggest we pursue establishing such a standard separately (I suggested something along those lines in [JuliaObjects/ConstructionBase.jl#54](https://github.com/JuliaObjects/ConstructionBase.jl/pull/54#issuecomment-1088765017)) and then use it in `get_compute_unit_impl` when available (even fewer types would need to specialize `get_compute_unit_impl` then).

@maleadt raised the question whether the compute unit concept belongs into Adapt. I strongly feel that it needs to be in a very lightweight package (so that generic code can be aware of compute devices without heavy deps, i.e. GPUArrays.jl is definitely too heavy for it). I would argue that Adapt is a good place, as `adapt(::AbstractComputeUnit, x)` is part of it. We could create a package ComputeUnits.jl, but Adapt.jl would have to depend on it. My suggestion would be to add the compute unit concept to Adapt.jl - we can still split off a ComputeUnits.jl in the future if necessary.

vchuravy commented on src/computedevs.jl:
```diff
@@ -0,0 +1,160 @@
+    UnknownComputeUnitOf(x) 
+end
+
+@generated function get_compute_unit_impl(::Type{TypeHistory}, x) where TypeHistory
```
I am really not a fan of this `@generated` implementation. I think it would be preferable to follow the Adapt.jl pattern here and to perform a tree-walk. Now it is probably the case that the current infrastructure is not general enough, but if this goes into Adapt.jl we should have one mechanism to do this.

@maleadt and I briefly spoke about this and in general we are not opposed to this functionality being in Adapt.jl

oschulz replied:
> @maleadt and I briefly spoke about this and in general we are not opposed to this functionality being in Adapt.jl

Thanks!

oschulz replied:
> I am really not a fan of this @generated implementation [...] would be preferable to follow the Adapt.jl pattern here and to perform a tree-walk [...] if this goes into Adapt.jl we should have one mechanism to do this.

Do you mean that it's `@generated`, or what it does? In principle it does a tree walk, I would say (and object's encountered on the way can use a different method/specialization of `get_compute_unit_impl`. The function could be written without `@generated`, it could be done with just `getfield`, `map` and `ntuple`. I only wrote it this way to minimize the resulting code and increase type stability. But maybe I'm missing your point here?

vchuravy replied:
There are two points:

1. I personally aim to reduce reliance on `@generated` as much as possible. So if we can write this without staged functions that would be great, it often makes the intent clearer and simplifies extendability (and is better for compile times, also better type-stability)
2. This duplicates the core functionality of Adapt.jl which is essentially a tree walk over structs, so instead of two implementations it would be great to have one, but Tim and I acknowledge that this might be a trickier design.

oschulz replied:
Ok, I'll try to rewrite without `@generated`.

oschulz replied:
> without staged functions [...] better type-stability

Type stability was actually one of the reasons I went for `@generated` here. :-)

Ok I tried without `@generated` @vchuravy, but type stability suffers. One implementation I came up with is

```
function get_compute_unit_impl_nogen(::Type{TypeHistory}, x::T) where {TypeHistory,T}
    NewTypeHistory = Union{TypeHistory, T}
    fields = ntuple(Base.Fix1(getfield, x), Val(fieldcount(T)))
    merge_compute_units(map(Base.Fix1(get_compute_unit_impl, NewTypeHistory), fields)...)
end
```

Nice and short, but:

```julia
julia> x = (a = 4.2, b = (c = rand(2,3)))
(a = 4.2, b = [0.7927795137326867 0.8930673224466184 0.15921059563423712; 0.6399176439174568 0.4501022168243579 0.3951239506670382])

julia> @inferred get_compute_unit_impl(Union{}, x)
Adapt.CPUDevice()

julia> @inferred get_compute_unit_impl_nogen(Union{}, x)
ERROR: return type Adapt.CPUDevice does not match inferred return type Any
```

And in truth, this version also uses generated code underneath, because `ntuple` does. Is there a way to do this without `ntuple`, efficiently?

oschulz commented:
@ChrisRackauckas how do you think this should be integrated with `ArrayInterfaceCore: AbstractDevice, AbstractCPU`?

ChrisRackauckas commented:
I'm not sure. @Tokazama might have opinions.

oschulz commented:
> This seems a lot like what we do in ArrayInterfaceCore.

Is this already in active use somewhere? I think we should definitely merge this in one lightweight place (ArrayInterfaceCore or Adapt) so that we establish a ecosystem-wide concept of a computing device/node.

Tokazama commented:
As you can see [here](https://github.com/JuliaArrays/ArrayInterface.jl/blob/b1ce7bb443eb82fc281be2989ee2fa5e364fe1d9/lib/ArrayInterfaceCore/src/ArrayInterfaceCore.jl#L354), we could do a better job of handling the actual device types in terms of documentation and additional methods. But we have a pretty good system for deriving the device type [here](https://github.com/JuliaArrays/ArrayInterface.jl/blob/b1ce7bb443eb82fc281be2989ee2fa5e364fe1d9/src/ArrayInterface.jl#L119).

Tokazama commented:
We have some nice tricks for avoiding generated methods using Static.jl too BTW. I haven't taken time to really dig in here and try it out, but it's worth thinking about.

oschulz commented:
> But we have a pretty good system for deriving the device type

Sure! Still, explicit support from the GPU packages would be nicer, right? And it's the only way that allows the user to handle systems with different GPUs cleanly, and also the only really clean way to bring KernelAbstractions into the mix as well.

> We have some nice tricks for avoiding generated methods using Static.jl too BTW.

Nice! Could that be used to make the structural fold in this PR type stable without generated code (and how)?

oschulz commented:
However we do this I think we should really aim to establish a single type tree for computing devices across the ecosystem, one that reaches down to different GPU (TPU, ...) types. And I think it would be very beneficial for users if they can write `adapt(some_device, some_ml_model_or_other_big_thing)`.

So we'll need a depenency between Adapt and ArrayInterfaceCore - the question is, which way round? Package maintainers, over to you. :-)

Tokazama commented:
> Sure! Still, explicit support from the GPU packages would be nicer, right? And it's the only way that allows the user to handle systems with different GPUs cleanly, and also the only really clean way to bring KernelAbstractions into the mix as well.

Explicit support for specific GPU devices is something we definitely want. It just takes buy in from those packages, which up until now was hard to get because we had so much latency. We still need to work out all the bugs to update to Static.jl v0.7 (which fixes a bunch of invalidations), but we're already at a load time of ~0.06 seconds on Julia v1.7.

> Nice! Could that be used to make the structural fold in this PR type stable without generated code (and how)?

From an initial look I'd say that `StaticSymbol` uses an internal generated function to merge `Symbols`. I'm pretty sure there's no other way to safely combine symbols that's also type stable.

Tokazama commented:
> However we do this I think we should really aim to establish a single type tree for computing devices across the ecosystem, one that reaches down to different GPU (TPU, ...) types. And I think it would be very beneficial for users if they can write `adapt(some_device, some_ml_model_or_other_big_thing)`.
> 
> So we'll need a depenency between Adapt and ArrayInterfaceCore - the question is, which way round? Package maintainers, over to you. :-)

Those maintaining GPU packages will know better if some approach has quirks that will work with their internals, so I'm just trying to explain the advantages of what we've developed thus far. The main advantage is we work pretty hard to navigate type wrappers in the most robust way we can so that if you grab a pointer from something you also can figure out what indices/strides/etc are valid.

oschulz commented:
> The main advantage is we work pretty hard to navigate type wrappers in the most robust way we can so that if you grab a pointer from something you also can figure out what indices/strides/etc are valid.

Nice! I guess we'll have to combine that with the Adapt-style "structural descend", since we don't just want to cover arrays, but also deeply nested objects in general (MC and statistics model, complex data and so on)?

I guess the question now is will Adapt depend on ArrayInterfaceCore or the other way round. Both have a load time under 1 ms now, so from a user perspective it won't matter much, and both are pretty much unavoidable dependencies in any real project anyway. Depending on preferences of the Adapt and ArrayInterface maintainers I can then try to merge this PR with what's in AI, either here or there.

oschulz commented:
> @Tokazama :  ArrayInterfaceCore. We don't have anything that merges device types and we account for a difference in CPU types that are built on different memory objects like a tuple

The equivalent to `ArrayInterface.CPUTuple` would be `Adapt.ComputingDeviceIndependent` in this PR (just by looking at, e.g., a StaticArray we can't know if we're currently on a CPU or GPU without additional context, and bitstypes don't really have to be "adapted" between computing devices).

oschulz commented:
I'd like to get this moving again - @Tokazama can `ArrayInterfaceCore.AbstractDevice` evolve/change a bit as part of this effort?

Tokazama commented:
I think the GPU side of things might be able to change without any breaking changes for us. We could probably justify some straightforward breaking changes if absolutely necessary if it unified the LoopVectorization and GPU ecosystems.

oschulz commented:
Thanks @Tokazama . Ok I'll take a look - otherwise we can just provide a conversion between the device API in this PR and `ArrayInterfaceCore.AbstractDevice`.

mcabbott commented:
We discussed this a little in the context of trying to make Flux.jl not depend on CUDA.jl

Some of the proposed benefits are already available from the very lightweight GPUArraysCore.jl. You can dispatch on `AbstractGPUArray`, and call `adapt` to move things. (This did not exist when the issue was opened.)

I don't think you can "query total and available memory". But perhaps that could be something like `GPUArraysCore.total_memory(::Type{<:Array}) = Sys.total_memory()` with a method for the type `CuArray`, instead of needing a new concept?

One thing you can't do is [`CUDA.functional`](https://cuda.juliagpu.org/stable/api/essentials/#CUDA.functional-Tuple{Bool}). Flux uses this so that calling `model |> gpu` does nothing if there's no GPU. And ideally it would like to know "what's the best compute unit around?"; getting back types like `Array` or `CuArray` would probably be perfect. (Getting back some device code would then need further layers before calling `adapt`.)

Tokazama commented:
It would be nice if we could have the GPU ecosystem and CPU/SIMD/LoopVectorization ecosystems share a common interface though.

oschulz commented:
Also, this PR is not just about checking if an array is a GPU array, but about checking what kind of device(s) a whole data structure is located on. Plus the ability to use adapt to move data structures to a specified device. I'm very happy about `GPUArraysCore`, but I think we definitely need a common super-type for accelerator *devices* as well as for arrays. KernelAbstractions.jl, for example, doesn't need to define an extra device type for each supported backend then.

As for pushing this forward: Maybe for now we can just work on fixing this up and merging this, and then look at possible connections to ArrayInterfaceCore.jl?

ToucheSir commented:
Since the pending review comments appear to be about tree traversal, would it be possible to try landing a reduced version of this PR with just `Adapt.adapt_storage(dev, x)`? _Maybe_ a constrained `get_computing_device(x::Union{Array,AbstractGPUArray})::AbstractComputingDevice` as a stretch goal. The former would be especially helpful on the Flux side.

oschulz commented:
> would it be possible to try landing a reduced version of this PR with just Adapt.adapt_storage(dev, x)? 

Fine with me - we can do the tree traversal in a second step. Maintainers, what do you think?

maleadt commented:
I'm still skeptical (this wasn't the scope of Adapt.jl and I feel like functionality is being tacked on because it's a common lightweight dependency), but if people are waiting for something like this I don't want to hold it up any longer 🙂  I guess it's fine to extend the scope to what the package is most commonly used for anyway. With respect to the actual functionality, we can always start with a getfield-based `get_computing_device` and try to figure out a better system to unify it with `adapt`'s tree traversal later.

oschulz commented:
> this wasn't the scope of Adapt.jl 

I do get that point, though maybe one justification would be that the current draft also specialized `adapt()` for devices?

> With respect to the actual functionality, we can always start with a getfield-based get_computing_device

Would the current recursive `get_compute_unit` in this draft fall under "getfield-based", or should the recursion/traversal be removed?

maleadt commented:
> Would the current recursive `get_compute_unit` in this draft fall under "getfield-based", or should the recursion/traversal be removed?

It's getfield-based as it recurses based on the struct's fields, as opposed to the current "rule"-based approach that Adapt takes (with its plenty `adapt_structure` methods). But that's fine for now, if we even want to change it, because the API doesn't restrict it to be getfield-based.

@vchuravy any final thoughts?

Tokazama commented:
I'm not sure if it makes a difference, but all of the device methods have been moved to ArrayInterfaceCore so if there are any reservations regarding integrating this because of start up latency, those should be mostly resolved.

oschulz commented:
It would be nice not to have divergent APIs here for Adapt and ArrayInterfaceCore - but `Adapt.adapt_storage(devive, obj)` (very useful on the user side I believe) could only be implemented if Adapt would depend on ArrayInterfaceCore (or the other way round). Could I get a quick maintainer opinion on whether that would be acceptable?

maleadt commented:
I see no reason not to; it's a dependency with minimal load time and no compatibility constraints.

Tokazama commented:
Also, load time is likely to stay that way. Since we got rid of requires I've been testing load time for nearly every PR to ArrayInterfaceCore. Probably should work that into automated testing for each PR at some point though...

oschulz commented:
> @maleadt: I see no reason not to; it's a dependency with minimal load time and no compatibility constraints.
> @Tokazama: Also, load time is likely to stay that way.

Ok, great - I'll redraft this PR-complex then, including ArrayInterfaceCore. I'm on holiday, gimme a bit of time ..

Jutho commented:
Bit late to the party, but if two packages (Adapt and ArrayInterfaceCore) would need this, and there is a discussion over which of the two should depend on the other, then why not have this functionality in a stand-alone package as was contemplated at the early stage of this PR?

oschulz commented:
Hi,

sorry for the long silence, been a bit overloaded with other stuff, but this is most definitely still on my to-do list. In fact, I've recently come across some use cases that really need something like `adapt(computing_device, object)`.

> why not have this functionality in a stand-alone package

we could, but as I understand there's no objection to Adapt depending on ArrayInterfaceCore, and since both are already very lightweight it might be overkill to introduce a third package for this. We can still do so later on if we need to, after all.

Tokazama commented:
> Bit late to the party, but if two packages (Adapt and ArrayInterfaceCore) would need this, and there is a discussion over which of the two should depend on the other, then why not have this functionality in a stand-alone package as was contemplated at the early stage of this PR?

It's not so much that two packages need this as much that one package already has this partially implemented

Jutho commented:
Ok, it was a question out of curiosity. Knowing what device an array is living on is a very basic primitive. For example, I have some custom implementations for `DenseArray` types, but these should only be used for host / CPU stuff, not GPU arrays. I am not particularly opposed to  depending `ArrayInterfaceCore`, but there is also quite a bit of stuff in there that I definitely do not need and seems perhaps SciML / Linsolve.jl specific (`lu_instance`), and some things that I don't really understand from the documentation what it is about and is probably matrix specific (`matrix_colors`?)

Certainly, at the moment, `ArrayInterfaceCore.device(::CuArray)` is producing the wrong result, so I will anyway go with depending on `GPUArraysCore` and using the `AbstractGPUArray` type.

Tokazama commented:
I'm not saying it can't be a third package. I just want to be clear that there's a bit more work there than one might assume because we need to avoid disrupting things already in production

oschulz commented:
Just a live sign, I haven't forgotten about this one. Due to recent changes in the interdependence between KernelAbstractions and the GPU packages it will have to evolve a bit though, I think. I'll play around with a few ideas.

vchuravy commented:
@oschulz also thank you for all the thinking about this. My long-term plan is to add KA as a dependency to GPUArrays, and anchor the notion of the compute device in KernelAbstractions e.g. as the backend.

I also took some inspiration and added an `allocate` function.

oschulz commented:
> My long-term plan is to add KA as a dependency to GPUArrays, and anchor the notion of the compute device in KernelAbstractions e.g. as the backend.

Thanks, that sounds great!

oschulz commented:
I'll explore the concepts in here in a standalone package [HeterogeneousComputing.jl](https://github.com/oschulz/HeterogeneousComputing.jl) for a while. I have a current need to use this kind of functionality in two or three real-life applications (so I'll register HeterogeneousComputing), and at least one of them will involve determining if data is on disk (DiskArrays/HDF) in addition to distinguishing just between CPU and GPU types. So I think I'll need to experiment and iterate this a bit more, before proposing a new incarnation of this in packages as central as Adapt and ArrayInterfaceCore.

Experimentation will be easier in a package that can go through versions faster without affecting half the ecosystem. :-) And with Julia v1.9 weakdeps is easier now without a huge Requires load-time hit.

I'll close this and open a "reminder issue" - I still believe it would be great to have common supertypes for compute units, adapt integration for them and a central way of determining what (nested) data is on which device.
